Move channel drop policy off of nano::socket and remove channel concurrency value as redundant

clemahieu opened the pull request:
This commit moves the responsibility for limiting writes to a socket from nano::socket on to nano::channel_tcp. The intended behavior of nano::socket is to be a thin wrapper around a raw tcp socket with the addition of queueing on a strand plus statistics gathering.
The drop policy is part of the channel abstraction which varies depending on the channel used. channel_udp uses the operating system to drop writes it can't service and channel_tcp needs to track the number of outstanding writes to determine what should be dropped.
The drop policy was being passed all the way through to nano::socket, punching a hole through the nano::socket abstraction.

This commit also removes the concurrency enum as redundant. All socket operations were being serialized through the strand implicitly and the only other function of having a non-concurrent channel was to bypass the socket rate limiter. Non-concurrent users of channel_tcp should manage their own rate limiting if they don't want their buffers dropped. The only users of non-concurrent channels appear to be bootstrapping and this limitation already appears to be respected.

This also removes the need for queueing the buffer and callback in a separate container as they're already captured in objects to be executed on the socket's strand. This removes the need to flush the buffers on socket close.

guilhermelawless commented on nano/node/socket.hpp:
```diff
@@ -84,25 +76,23 @@ class socket : public std::enable_shared_from_this<nano::socket>
-	void flush_send_queue_callbacks ();
+
+public:
+	size_t const queue_size_max = 128;
```
This could be `static` and `constexpr`

cryptocode commented:
> This commit also removes the concurrency enum as redundant.

The purpose was to avoid enqueuing bootstrap writes. In that case it was `async_write->handler->async_write->...` (such as `bulk_pull_server::send_next`), while now it'll be `post (strand queuing)->async_write->handler->post->async_write->...`

Not sure if it'll affect performance/memory, just wanted to point it out since it's a significant amount of calls/data being processed.

clemahieu commented:
> es. In that case it was `async_write->handler->async_write->...` (such as `bulk_pull_server::send_next`), while now it'll be `post (stran`

Good point. I'm working on a refactor of the bootstrapping code, hopefully using coroutines. I'll keep that in mind to profile.

clemahieu committed:
This commit moves the responsibility for limiting writes to a socket from nano::socket on to nano::channel_tcp. The intended behavior of nano::socket is to be a thin wrapper around a raw tcp socket with the addition of queueing on a strand plus statistics gathering.
The drop policy is part of the channel abstraction which varies depending on the channel used. channel_udp uses the operating system to drop writes it can't service and channel_tcp needs to track the number of outstanding writes to determine what should be dropped.
The drop policy was being passed all the way through to nano::socket, punching a hole through the nano::socket abstraction.

This commit also removes the concurrency enum as redundant. All socket operations were being serialized through the strand implicitly and the only other function of having a non-concurrent channel was to bypass the socket rate limiter. Non-concurrent users of channel_tcp should manage their own rate limiting if they don't want their buffers dropped. The only users of non-concurrent channels appear to be bootstrapping and this limitation already appears to be respected.

This also removes the need for queueing the buffer and callback in a separate container as they're already captured in objects to be executed on the socket's strand. This removes the need to flush the buffers on socket close.

SergiySW commented on nano/core_test/socket.cpp:
```diff
@@ -41,22 +37,25 @@ TEST (socket, drop_policy)
+		ASSERT_EQ (1, client.use_count ());
 	};
 
 	func (max_write_queue_size * 2 + 1, nano::buffer_drop_policy::no_socket_drop);
```
Can it use `nano::socket::queue_size_max` directly? It seems `max_write_queue_size `is not used in other places

guilhermelawless reviewed (approved):
TSAN and ASAN passing, tested a short bootstrap on Debug also. Great change!
